kernel/src/process_checking.rs

```diff
@@ -0,0 +1,376 @@
+/// processes with the same Application Identifier cannot run concurrently.
+pub trait AppUniqueness {
+    /// Returns whether `process_a` and `process_b` have a different identifier,
+    /// and so can run concurrently. If this returns `false`, the kernel
+    /// will not run `process_a` and `process_b` at the same time.
+    fn different_identifier(&self, _process_a: &dyn Process, _process_b: &dyn Process) -> bool {
+        true
+    }
+
+    /// Return whether there is a currently running process that has
+    /// the same application identifier as `process`. This means that
+    /// if `process` is currently running, `has_unique_identifier`
+    /// returns false.
+    fn has_unique_identifier(
+        &self,
+        process: &dyn Process,
+        processes: &[Option<&dyn Process>],
+    ) -> bool {
+        let len = processes.len();
+        // If the process is running or not runnable it does not have a unique identifier;
+        // these two states describe a process that is potentially runnable, dependent on
+        // checking for identifier uniqueness at runtime.
+        if process.get_state() != State::CredentialsApproved
+            && process.get_state() != State::Terminated
+        {
+            return false;
+        }
+
+        // Note that this causes `process` to compare against itself;
+        // however, since `process` should not be running, it will
+        // not check the identifiers and say they are different. This means
+        // this method returns false if the process is running.
+        for i in 0..len {
+            let checked_process = processes[i];
+            let diff = checked_process.map_or(true, |other| {
+                !other.is_running() || self.different_identifier(process, other)
+            });
+            if !diff {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+/// Default implementation.
+impl AppUniqueness for () {}
+
+/// Transforms Application Credentials into a corresponding ShortID.
+pub trait Compress {
+    fn to_short_id(&self, _credentials: &TbfFooterV2Credentials) -> ShortID {
+        ShortID::LocallyUnique
+    }
+}
+
+impl Compress for () {}
+
+pub trait CredentialsCheckingPolicy<'a>:
+    AppCredentialsChecker<'a> + Compress + AppUniqueness
+{
+}
+impl<'a, T: AppCredentialsChecker<'a> + Compress + AppUniqueness> CredentialsCheckingPolicy<'a>
+    for T
+{
+}
+
+/// A sample Credentials Checking Policy that loads and runs Userspace
+/// Binaries with unique process names; if it encounters a Userspace
+/// Binary with the same process name as an existing one it fails the
+/// uniqueness check and is not run.
+pub struct AppCheckerSimulated<'a> {
+    deferred_caller: &'a DynamicDeferredCall,
+    handle: OptionalCell<DeferredCallHandle>,
+    client: OptionalCell<&'a dyn Client<'a>>,
+    credentials: OptionalCell<TbfFooterV2Credentials>,
+    binary: OptionalCell<&'a [u8]>,
+}
```

Comment:
It might be a good idea to do what we did with schedulers and move these to their own files in a new module. It's already a little difficult to discover and document the available checkers.

Reply:
OK, this makes sense now that we have an implementation on (). Given how simple they are I am on one hand tempted to just keep them in a file (`process_checkers.rs`) for now, rather than a new module. But given that there will almost certainly be new ones, cleaning up the namespace now through a new module and planning for that sounds better.

Reply:
Done.